Gluster Operator Design

jarrpa opened the pull request:
A design doc to hopefully get some traction on the development of a Gluster operator. From the introduction:

```
This document aims to settle on a concrete set of functionality that a Gluster
operator should have. It is outside the scope of this document to specify any
implementation details of this functionality. The desired functionality should
not be strictly limited by the extents of current-day technologies but should
not dismiss the costs of waiting for and executing forward development in those
technologies.

In addition, this document only specifies the desired functionality for
declaring a 1.0 release. Further enhancements and feature roadmaps will be
captured elsewhere.
```

Signed-off-by: Jose A. Rivera <jarrpa@redhat.com>

phlogistonjohn commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+## Configurations
+
+The Gluster operator will be capable of managing a number of configurations of
+Gluster clusters. Gluster's flexibility allows it to run in many typs of
```
Minor typo

vbellur commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+
+The Gluster operator will monitor the health of all volumes in an automated
+Gluster cluster. It will report a "Warning" when a volume reports failing bricks
+and will report a "Critical" problem when a volume becomes inaccessible.
```
It would be nice to monitor essential (glusterd, self-heal-daemon etc.) & optional services (quotad, bitrot-daemon) in addition to bricks as part of this activity.

sbarre-esit replied:
Besides failing bricks, it would be good to also check for unhealed files, or files in split-brain.

JohnStrunk commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+
+**Target Configurations:** Managed
+
+#### Cluster Scaling
```
I think this should be out of scope at this time. GD2 should fix the need for multiple clusters, meaning we'd only need this if it falls short... a good reason to push it out.

JohnStrunk commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+take any number of corrective actions to recover from a situation that could
+result in potential data loss.
+
+#### Brick Replacement
```
Per the meeting last month, this will be a Gluster-internal operation, assuming sufficient free capacity is available. The operator is only concerned with adding and removing capacity & nodes from the cluster.

jarrpa replied:
Oh, right! Fair enough. Tying this back to the monitoring section, would it make sense to have some sort of "out of space" event that the operator could react to in addition to the two watermarks?

JohnStrunk commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Device Replacement
```
As a first step, this can be greatly simplified by having 1 device per gluster node, making a failed device the same as a failed node.
That 1 device could be an arbitrarily sized RAID array if an admin chooses.

JohnStrunk commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+
+**Target Configurations:** Managed
+
+#### Node Replacement
```
Here again, there is a split between the operator's tasks and Gluster's internal responsibilities. The operator would create a new node, and once data has been rebuilt, destroy the old node. Gluster would internally handle the brick reconstruction.

mrobson commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Cluster Upgrade
```
Can the cluster upgrade component be made dynamic enough to also be useful in other situations which require a rolling restart of the Gluster nodes? Effectively, it would handle bringing down Gluster cleanly, waiting for X and ensuring the health before continuing. Beyond the Gluster upgrade, the 2 primary examples of X being, upgrading the underlying nodes and restarting the VMs or Servers for OS Patching.

jarrpa replied:
The operator would not be managing the upgrade of any underlying infrastructure. In the case where a given node identity can not move between hosts, it would be up to the administrator to ensure they upgrade such hosts one at a time and wait for the Gluster cluster's health to restore. The only involvement of the operator in that case would be, at best, to report the state of health.

mrobson replied:
Agreed, it would not handle those other tasks, but ideally, it would be able to aid them. Instead of everyone else needing to understand how to cleanly terminate gluster (daemon set / pod or otherwise) and check it's health when it comes back, it would be nice for other users and operators to be able to leverage the gluster operator for that functionality. If the tasks are broken out as discretely, that may be possible. Long and short, that kind of reusability is what I am hoping you take into consideration.

jarrpa replied:
At present, and best I can see, there's nothing more to bringing down a Gluster pod than any other pod. Kubernetes will take care to do as clean a shutdown as it can, to my understanding. As far as watching to health restoration, would simply exposing the state via a condition field in an object's Status be sufficient?

jarrpa replied:
> There is something more in that the volume with bricks hosted by that pod should be clean before shutting down. I'm not sure if sig lifecycle would have any opinions on how that could be ensured.

That's up to Gluster, not the operator. Kube sends a SIGTERM to the main process on pod delete, in our case some sort of init/systemd, which should take care of shutting things down gracefully.

jarrpa commented:
> You might consider adding the CRD spec to the design doc. This will help define the desired state the admin can specify when he creates the gluster cluster. It will also help differentiate between what is desired state managed by the operator vs configuration tasks that the operator likely wouldn't manage.

I would think the CRD spec would be too close to the implementation of the operator to merit including in this document. As it stands, the CRD in my prototype has evolved over the course of the implementation as I think things through and try things out.

> For the configuration tasks, consider that gluster should have a "toolbox" pod that allows the admin to drop in and perform operational tasks manually. The toolbox allows running tasks that the operator might not be expected to run, either because they are one-time admin tasks or because the admin just needs to override the operator.

Wouldn't that just be the same as exec'ing into one of the Gluster nodes?

jarrpa commented:
@travisn Ah, okay. The desired state is only changed by the operator in the case of a managed deployment, where both the services and the storage devices are not tied to specific hosts. Otherwise it is up to the administrator to change the CRD object.

I can see the rational for some general state outline, at least. I'll work on that.

phlogistonjohn commented:
Perhaps this is too implementation oriented but I think it would be nice if the ability to make changes to the cluster were individually toggle-able. For example you are working with a "managed" cluster but do not want it to automatically perform Cluster scaling but storage scaling might be ok.

IOW, are the target configurations in the doc more like use cases or something that would actually be specified in the system configuration?

sbarre-esit commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+
+The Gluster operator will monitor the health of all nodes in an automated
+Gluster cluster. It will report a "Critical" problem when a node becomes
+inaccessible.
```
I'll second @vbellur request to monitor additional services. Currently the readiness probe only checks `glusterd`, while `tcmu-runner` `gluster-block-target` and `gluster-blockd` are all critical to a healthy gluster pod. Perhaps that just needs to be pushed into the readiness probe though?

sbarre-esit replied:
Would this also include the other pods like `glusterblock-provisioner` and `heketi` ?

jarrpa replied:
> Perhaps that just needs to be pushed into the readiness probe though?

Yes. :)

> Would this also include the other pods like `glusterblock-provisioner` and `heketi` ?

That's implementation-specific. Neither of those two are guaranteed to always be around.

obnoxxx replied:
> >  Would this also include the other pods like glusterblock-provisioner and heketi ?
> 
> That's implementation-specific. Neither of those two are guaranteed to always be around.

But I think in the current situation, they would certainly belong to the scope of the operator!

obnoxxx replied:
I think this point is valid though. The document reads as if the operator only cares about gluster nodes. But maybe we need to define what are the "nodes in an automated Gluster cluster". Those will be gluster nodes, but possibly other pods in kubernetes. At least currently it will include heketi and (optionally) a gluster-block-provisioner pod. Etc.

jarrpa commented:
Updated the PR:

* Added a description of state information that the operator will maintain.
* Removed the "cluster scaling" feature.

jarrpa commented on docs/design/operator.md:
```diff
@@ -77,15 +77,7 @@ possible to allow for as minimal a configuration as possible.
-stage of initialization. The operator's desired configuration for the cluster
-will need to be compatible with the state of the cluster at the time of import.
-
-**Target Configurations:** External, Hosted, Managed
```
Naw, I just won't hold up a 1.0 for it. It'll either be in a 1.5ish release or developed as an external tool to migrate existing clusters to operator configurations.

humblec commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+
+This document specifies the overall design for a Gluster operator. An operator
+is a Kubernetes [controller](https://github.com/kubernetes/sample-controller)
+that is focused on managing a specific application with the goal of automating
```
I dont think we have to specify the `sample controller` here, the reason being no bits about implementation is covered here in the PR. We have to discuss further on the implementation side and come up best suited controller model, there are different framework available to implement an operator ( meta controller , operator sdk, ..etc),  we have to discuss further on it and come up with an implementation doc or PR later.

jarrpa replied:
Find me a better link to showcase what a controller is and I'll happily change it. :)

jarrpa replied:
I've asked around about the operator SDK and it very much doesn't even exist yet. While it sounds like it might be useful, I don't feel like waiting around for it.

Reading on metacontroller, it looks like a utility API for writing controllers. It doesn't sound like it has anything particularly novel, and in some aspects looks more constricting than just writing the controller framework by hand. Given that I've already written a controller base, I'm not interested in exploring metacontroller at this time.

JohnStrunk replied:
> I've asked around about the operator SDK and it very much doesn't even exist yet. While it sounds like it might be useful, I don't feel like waiting around for it.

The SDK is here: https://github.com/operator-framework/operator-sdk
While it is still very actively changing, it is certainly worth looking at as a template. It is also probably going to be useful to be compatible with the lifecycle manager: https://github.com/operator-framework/operator-lifecycle-manager

jarrpa replied:
Okay, more accurately: I was told not to use it yet. :) It effectively does not exist for me until it at least hits alpha.

obnoxxx commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+# Gluster Operator 1.0 Design
```
Not sure if 1.0 is really appropriate here. It sounds premature and might set false expectations.

obnoxxx commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+technologies.
+
+In addition, this document only specifies the desired functionality for
+declaring a 1.0 release. Further enhancements and feature roadmaps will be
```
I'd rather say something along the lines of "...specifies a desired minimal set of functionalities for getting started with an operator" ... (again the 1.0 comment).

obnoxxx commented on docs/design/operator.md:
```diff
@@ -0,0 +1,200 @@
+#### Device Health
+
+The Gluster operator will monitor the health of all devices in an automated
+Gluster cluster. It will report a "Warning" when a device reports failing health
```
I think we need to define the term "automated Gluster cluster" in this document. (see below comments)

jarrpa commented:
Pushed updates:

* Changed desired version from `1.0` to `0.1`.
* Removed the notion of "automated Gluster clusters", I already assume elsewhere that when I say "Gluster cluster" the reader knows I'm talking about those Gluster clusters that are deployed and automated by the operator.
